{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Infer EC50 values for the `171201_Eva` chip"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import `Python` modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import re\n",
    "import glob\n",
    "\n",
    "import xml.etree.ElementTree as ET\n",
    "sys.path.append(\"/home/jupyter/tacc-work/jupyter_packages/lib/python2.7/site-packages\")\n",
    "from FlowCytometryTools import *\n",
    "\n",
    "import numpy as np\n",
    "import pandas\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "sns.set(font_scale=1.5)\n",
    "\n",
    "# Initialize directories\n",
    "resultsdir = \"results/171201_Eva_10861\"\n",
    "resultsdir_parse_metadata = os.path.join(resultsdir, 'parse_metadata')\n",
    "dirs = [resultsdir, resultsdir_parse_metadata]\n",
    "for dir_i in dirs:\n",
    "    if not os.path.isdir(dir_i):\n",
    "        os.makedirs(dir_i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make the input experimental-summary file\n",
    "\n",
    "This library has multiple replicates. Below is a list of replicates, as specified by unique `plan_id` variables from the BIOFAB, along with the matched and naive controls, which lack `plan_id`s. Devin said that the `aq_item_id` for the naive sample for the replicate with the lower `plan_id` also has a lower `aq_item_id` than the other naive sample.\n",
    "\n",
    "* `plan_id`: 10861; `aq_item_id` for matched naive control: 128412\n",
    "* `plan_id`: 11284; `aq_item_id` for matched naive control: 128440\n",
    "\n",
    "### Specify all inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>aq_item_id</th>\n",
       "      <th>plan_id</th>\n",
       "      <th>strain</th>\n",
       "      <th>fastq_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protease</th>\n",
       "      <th>concentration</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">chymotrypsin</th>\n",
       "      <th>9.5</th>\n",
       "      <td>128444</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128444/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28.0</th>\n",
       "      <td>128446</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128446/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83.0</th>\n",
       "      <td>128456</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128456/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250.0</th>\n",
       "      <td>128452</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128452/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750.0</th>\n",
       "      <td>128460</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128460/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2250.0</th>\n",
       "      <td>128462</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128462/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">trypsin</th>\n",
       "      <th>0.0</th>\n",
       "      <td>128412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128412/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9.5</th>\n",
       "      <td>128448</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128448/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28.0</th>\n",
       "      <td>128450</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128450/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83.0</th>\n",
       "      <td>128454</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128454/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250.0</th>\n",
       "      <td>128458</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128458/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750.0</th>\n",
       "      <td>128464</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128464/Files/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2250.0</th>\n",
       "      <td>128466</td>\n",
       "      <td>10861.0</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>128466/Files/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            aq_item_id  plan_id     strain       fastq_id\n",
       "protease     concentration                                               \n",
       "chymotrypsin 9.5                128444  10861.0  EVA_NOV17  128444/Files/\n",
       "             28.0               128446  10861.0  EVA_NOV17  128446/Files/\n",
       "             83.0               128456  10861.0  EVA_NOV17  128456/Files/\n",
       "             250.0              128452  10861.0  EVA_NOV17  128452/Files/\n",
       "             750.0              128460  10861.0  EVA_NOV17  128460/Files/\n",
       "             2250.0             128462  10861.0  EVA_NOV17  128462/Files/\n",
       "trypsin      0.0                128412      NaN  EVA_NOV17  128412/Files/\n",
       "             9.5                128448  10861.0  EVA_NOV17  128448/Files/\n",
       "             28.0               128450  10861.0  EVA_NOV17  128450/Files/\n",
       "             83.0               128454  10861.0  EVA_NOV17  128454/Files/\n",
       "             250.0              128458  10861.0  EVA_NOV17  128458/Files/\n",
       "             750.0              128464  10861.0  EVA_NOV17  128464/Files/\n",
       "             2250.0             128466  10861.0  EVA_NOV17  128466/Files/"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify the location of FASTQ files\n",
    "fastq_dir = '/work/05402/haddox/jupyter/sd2e-community/shared-q1-workshop/strcklnd/ngs_data/run_8/EVA_Nov17_new_prep/'\n",
    "fastq_summary_file = '/work/05402/haddox/jupyter/sd2e-community/shared-q1-workshop/strcklnd/ngs_data/run_8/manifest.csv'\n",
    "\n",
    "# Specify the locations of directories with FCS files with part of the FACS data\n",
    "facs_dir = \"/work/05402/haddox/jupyter/sd2e-community/shared-q1-workshop/strcklnd/facs_data/production_data/\"\n",
    "job_ids = ['Job_60359', 'Job_60063', 'Job_59787']\n",
    "plan_id = 10861\n",
    "facs_dirs = ','.join([os.path.join(facs_dir, job_id) for job_id in job_ids])\n",
    "job_ids = ','.join(job_ids)\n",
    "\n",
    "# Specify the FACS channel name to investigate\n",
    "facs_channel_name = u'FITC-A'\n",
    "\n",
    "# Make a list of input XML files with the other part of the FACS data\n",
    "xml_files = ','.join(glob.glob(\n",
    "    os.path.join(facs_dir, 'sort_reports', '*/*.xml')\n",
    "))\n",
    "\n",
    "# Specify column-specific values to analyze in the metadata files.\n",
    "library_name = \"EVA_NOV17\"\n",
    "ignore_aq_item_ids = '128440,128442' # ignore the sequencing data for the naive sample for the other replicate and the (trypsin, 0) sample for this replicate\n",
    "\n",
    "# Specify the output metadata CSV file\n",
    "experimental_metadata_output_file = os.path.join(resultsdir_parse_metadata, 'experimental_metadata_from_script.csv')\n",
    "\n",
    "# Read in input arguments, as if it were a new script\n",
    "job_ids = job_ids.split(',')\n",
    "facs_dirs = facs_dirs.split(',')\n",
    "xml_files = xml_files.split(',')\n",
    "ignore_aq_item_ids = ignore_aq_item_ids.split(',')\n",
    "\n",
    "# Read in data on location of deep-sequencing data\n",
    "fastq_df = pandas.read_csv(fastq_summary_file)\n",
    "\n",
    "# Remove entires to ignore, as specified above\n",
    "indices_to_drop = [i for (i, row) in fastq_df.iterrows() if str(row['aq_item_id']) in ignore_aq_item_ids]\n",
    "fastq_df.drop(indices_to_drop, inplace=True)\n",
    "\n",
    "# Piece together paths to FASTQ files and tweak some of entries to match\n",
    "# expected patterns/cases\n",
    "fastq_df['fastq_id'] = fastq_df.apply(lambda row: os.path.join(\n",
    "    str(row['aq_item_id']), 'Files/'\n",
    "), axis=1)\n",
    "fastq_df['protease'] = fastq_df['protease'].apply(lambda x: x.lower())\n",
    "fastq_df['strain'] = fastq_df['strain'].apply(lambda x: x.replace(' ', '_'))\n",
    "\n",
    "# To conform with previous naming schemes, I will relabel the\n",
    "# naive sample (naive, 0) to be the naive trypsin sample (trypsin, 0)\n",
    "fastq_df['protease'] = fastq_df['protease'].apply(lambda x: x.replace('naive', 'trypsin'))\n",
    "\n",
    "# Index by protease name and concentration\n",
    "fastq_df.set_index(['protease', 'concentration'], inplace=True)\n",
    "\n",
    "# Downsample to include only samples within library of interest using the\n",
    "# `plan_id` to select the replicate of interest\n",
    "fastq_df = fastq_df[\n",
    "    (fastq_df['strain'] == library_name) &\\\n",
    "    (fastq_df['plan_id'].isin([plan_id, np.nan]))\n",
    "]\n",
    "fastq_df.sort_index(inplace=True)\n",
    "fastq_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For each sample in the experiment, quantify the total number of events and the number of events that passed the sorting threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def s_log(sample, channel_names):\n",
    "    new_sample = sample.copy()\n",
    "    new_data = new_sample.data\n",
    "\n",
    "    for channel_name in channel_names:\n",
    "        new_data[channel_name] = np.log10(new_data[channel_name])\n",
    "        \n",
    "    new_data.replace(to_replace = -np.inf, value = -1, inplace = True)\n",
    "    new_data = new_data.dropna()\n",
    "    new_sample.data = new_data\n",
    "    \n",
    "    return new_sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read in data from the `manifest.txt` files\n",
    "manifest_df = pandas.DataFrame()\n",
    "for (job_id, facs_dir) in zip(job_ids, facs_dirs):\n",
    "    df = pandas.read_csv(os.path.join(facs_dir, 'manifest.txt'))\n",
    "    df['job_id'] = job_id\n",
    "    df['directory'] = facs_dir\n",
    "    df['specimen'] = df['filename'].apply(lambda x : x[:12])\n",
    "    df['tube'] = df['filename'].apply(lambda x : x[13:21])\n",
    "    df['strain'] = df['strain'].apply(lambda x: x.replace(' ', '_'))\n",
    "    manifest_df = pandas.concat([manifest_df, df])\n",
    "manifest_df.set_index(['job_id', 'specimen', 'tube'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/python2/lib/python2.7/site-packages/ipykernel/__main__.py:6: RuntimeWarning: invalid value encountered in log10\n",
      "/opt/conda/envs/python2/lib/python2.7/site-packages/ipykernel/__main__.py:6: RuntimeWarning: divide by zero encountered in log10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>protease</th>\n",
       "      <th>concentration</th>\n",
       "      <th>total_number_of_recorded_events</th>\n",
       "      <th>number_of_events_passing_fitc_gate</th>\n",
       "      <th>fraction_collected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>job_id</th>\n",
       "      <th>specimen</th>\n",
       "      <th>tube</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Job_60359</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_010</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99993</td>\n",
       "      <td>77560</td>\n",
       "      <td>0.775654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>99996</td>\n",
       "      <td>21064</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>99995</td>\n",
       "      <td>6140</td>\n",
       "      <td>0.061403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_011</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>76259</td>\n",
       "      <td>0.762613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>19489</td>\n",
       "      <td>0.194896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>2217</td>\n",
       "      <td>0.022171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Job_60063</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_010</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99995</td>\n",
       "      <td>67805</td>\n",
       "      <td>0.678084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>98303</td>\n",
       "      <td>64106</td>\n",
       "      <td>0.652127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>99998</td>\n",
       "      <td>52457</td>\n",
       "      <td>0.524580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_011</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99995</td>\n",
       "      <td>70467</td>\n",
       "      <td>0.704705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>99993</td>\n",
       "      <td>64109</td>\n",
       "      <td>0.641135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>53023</td>\n",
       "      <td>0.530246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Job_59787</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>37520</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>99999</td>\n",
       "      <td>32048</td>\n",
       "      <td>0.320483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>100000</td>\n",
       "      <td>29746</td>\n",
       "      <td>0.297460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>99999</td>\n",
       "      <td>30930</td>\n",
       "      <td>0.309303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_010</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>99999</td>\n",
       "      <td>27904</td>\n",
       "      <td>0.279043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain      protease  concentration  \\\n",
       "job_id    specimen     tube                                               \n",
       "Job_60359 Specimen_001 Tube_010  EVA_NOV17  chymotrypsin            0.0   \n",
       "                       Tube_006  EVA_NOV17  chymotrypsin          750.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin         2250.0   \n",
       "                       Tube_011  EVA_NOV17       trypsin            0.0   \n",
       "                       Tube_008  EVA_NOV17       trypsin          750.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin         2250.0   \n",
       "Job_60063 Specimen_001 Tube_010  EVA_NOV17  chymotrypsin            0.0   \n",
       "                       Tube_006  EVA_NOV17  chymotrypsin           83.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin          250.0   \n",
       "                       Tube_011  EVA_NOV17       trypsin            0.0   \n",
       "                       Tube_008  EVA_NOV17       trypsin           83.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin          250.0   \n",
       "Job_59787 Specimen_001 Tube_006  EVA_NOV17       trypsin            0.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin            9.5   \n",
       "                       Tube_008  EVA_NOV17  chymotrypsin           28.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin            9.5   \n",
       "                       Tube_010  EVA_NOV17       trypsin           28.0   \n",
       "\n",
       "                                 total_number_of_recorded_events  \\\n",
       "job_id    specimen     tube                                        \n",
       "Job_60359 Specimen_001 Tube_010                            99993   \n",
       "                       Tube_006                            99996   \n",
       "                       Tube_007                            99995   \n",
       "                       Tube_011                            99997   \n",
       "                       Tube_008                            99997   \n",
       "                       Tube_009                            99997   \n",
       "Job_60063 Specimen_001 Tube_010                            99995   \n",
       "                       Tube_006                            98303   \n",
       "                       Tube_007                            99998   \n",
       "                       Tube_011                            99995   \n",
       "                       Tube_008                            99993   \n",
       "                       Tube_009                            99997   \n",
       "Job_59787 Specimen_001 Tube_006                            99997   \n",
       "                       Tube_007                            99999   \n",
       "                       Tube_008                           100000   \n",
       "                       Tube_009                            99999   \n",
       "                       Tube_010                            99999   \n",
       "\n",
       "                                 number_of_events_passing_fitc_gate  \\\n",
       "job_id    specimen     tube                                           \n",
       "Job_60359 Specimen_001 Tube_010                               77560   \n",
       "                       Tube_006                               21064   \n",
       "                       Tube_007                                6140   \n",
       "                       Tube_011                               76259   \n",
       "                       Tube_008                               19489   \n",
       "                       Tube_009                                2217   \n",
       "Job_60063 Specimen_001 Tube_010                               67805   \n",
       "                       Tube_006                               64106   \n",
       "                       Tube_007                               52457   \n",
       "                       Tube_011                               70467   \n",
       "                       Tube_008                               64109   \n",
       "                       Tube_009                               53023   \n",
       "Job_59787 Specimen_001 Tube_006                               37520   \n",
       "                       Tube_007                               32048   \n",
       "                       Tube_008                               29746   \n",
       "                       Tube_009                               30930   \n",
       "                       Tube_010                               27904   \n",
       "\n",
       "                                 fraction_collected  \n",
       "job_id    specimen     tube                          \n",
       "Job_60359 Specimen_001 Tube_010            0.775654  \n",
       "                       Tube_006            0.210648  \n",
       "                       Tube_007            0.061403  \n",
       "                       Tube_011            0.762613  \n",
       "                       Tube_008            0.194896  \n",
       "                       Tube_009            0.022171  \n",
       "Job_60063 Specimen_001 Tube_010            0.678084  \n",
       "                       Tube_006            0.652127  \n",
       "                       Tube_007            0.524580  \n",
       "                       Tube_011            0.704705  \n",
       "                       Tube_008            0.641135  \n",
       "                       Tube_009            0.530246  \n",
       "Job_59787 Specimen_001 Tube_006            0.375211  \n",
       "                       Tube_007            0.320483  \n",
       "                       Tube_008            0.297460  \n",
       "                       Tube_009            0.309303  \n",
       "                       Tube_010            0.279043  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Quantify the total number of events and events passing the gate\n",
    "fitc_data_dict = {}\n",
    "total_number_of_recorded_events = []\n",
    "number_of_events_passing_fitc_gate = []\n",
    "for (i, row) in manifest_df.iterrows():\n",
    "    \n",
    "    # Read in the data for the appropriate channel and transform it to a log10 scale\n",
    "    fcs_file = os.path.join(row['directory'], row['filename'])\n",
    "    sample = FCMeasurement(ID='Test Sample', datafile=fcs_file)\n",
    "    #print(\"Analyzing the channel: {0}\".format(channel_name))\n",
    "    sample = s_log(sample, [facs_channel_name])\n",
    "    df = sample.data\n",
    "    \n",
    "    # Count the number of total events and events passing the gate\n",
    "    total_number_of_recorded_events.append(len(df.index.values))\n",
    "    number_of_events_passing_fitc_gate.append(\n",
    "        sum(df[facs_channel_name] > 3.0)\n",
    "    )\n",
    "\n",
    "# Append the FACS data to the manifest dataframe\n",
    "manifest_df['total_number_of_recorded_events'] = total_number_of_recorded_events\n",
    "manifest_df['number_of_events_passing_fitc_gate'] = number_of_events_passing_fitc_gate\n",
    "manifest_df['fraction_collected'] = \\\n",
    "    manifest_df['number_of_events_passing_fitc_gate'] / manifest_df['total_number_of_recorded_events']\n",
    "columns_to_show = [\n",
    "    'strain', 'protease', 'concentration', 'total_number_of_recorded_events',\n",
    "    'number_of_events_passing_fitc_gate', 'fraction_collected'\n",
    "]\n",
    "manifest_df = manifest_df[\n",
    "    manifest_df['strain'] == library_name\n",
    "]\n",
    "manifest_df[columns_to_show]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Go through the dataframe, making a list of samples to drop. Specifically, I will drop naive samples that are not from the first day of sorting. But, before doing so, I will record data from them to include in the `parent_expression` column of the output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>strain</th>\n",
       "      <th>protease</th>\n",
       "      <th>concentration</th>\n",
       "      <th>total_number_of_recorded_events</th>\n",
       "      <th>number_of_events_passing_fitc_gate</th>\n",
       "      <th>fraction_collected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>job_id</th>\n",
       "      <th>specimen</th>\n",
       "      <th>tube</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Job_60359</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>99996</td>\n",
       "      <td>21064</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>99995</td>\n",
       "      <td>6140</td>\n",
       "      <td>0.061403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>19489</td>\n",
       "      <td>0.194896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>2217</td>\n",
       "      <td>0.022171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Job_60063</th>\n",
       "      <th rowspan=\"4\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>98303</td>\n",
       "      <td>64106</td>\n",
       "      <td>0.652127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>99998</td>\n",
       "      <td>52457</td>\n",
       "      <td>0.524580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>99993</td>\n",
       "      <td>64109</td>\n",
       "      <td>0.641135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>53023</td>\n",
       "      <td>0.530246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">Job_59787</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Specimen_001</th>\n",
       "      <th>Tube_006</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99997</td>\n",
       "      <td>37520</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_007</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>99999</td>\n",
       "      <td>32048</td>\n",
       "      <td>0.320483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_008</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>100000</td>\n",
       "      <td>29746</td>\n",
       "      <td>0.297460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_009</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>99999</td>\n",
       "      <td>30930</td>\n",
       "      <td>0.309303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tube_010</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>99999</td>\n",
       "      <td>27904</td>\n",
       "      <td>0.279043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    strain      protease  concentration  \\\n",
       "job_id    specimen     tube                                               \n",
       "Job_60359 Specimen_001 Tube_006  EVA_NOV17  chymotrypsin          750.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin         2250.0   \n",
       "                       Tube_008  EVA_NOV17       trypsin          750.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin         2250.0   \n",
       "Job_60063 Specimen_001 Tube_006  EVA_NOV17  chymotrypsin           83.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin          250.0   \n",
       "                       Tube_008  EVA_NOV17       trypsin           83.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin          250.0   \n",
       "Job_59787 Specimen_001 Tube_006  EVA_NOV17       trypsin            0.0   \n",
       "                       Tube_007  EVA_NOV17  chymotrypsin            9.5   \n",
       "                       Tube_008  EVA_NOV17  chymotrypsin           28.0   \n",
       "                       Tube_009  EVA_NOV17       trypsin            9.5   \n",
       "                       Tube_010  EVA_NOV17       trypsin           28.0   \n",
       "\n",
       "                                 total_number_of_recorded_events  \\\n",
       "job_id    specimen     tube                                        \n",
       "Job_60359 Specimen_001 Tube_006                            99996   \n",
       "                       Tube_007                            99995   \n",
       "                       Tube_008                            99997   \n",
       "                       Tube_009                            99997   \n",
       "Job_60063 Specimen_001 Tube_006                            98303   \n",
       "                       Tube_007                            99998   \n",
       "                       Tube_008                            99993   \n",
       "                       Tube_009                            99997   \n",
       "Job_59787 Specimen_001 Tube_006                            99997   \n",
       "                       Tube_007                            99999   \n",
       "                       Tube_008                           100000   \n",
       "                       Tube_009                            99999   \n",
       "                       Tube_010                            99999   \n",
       "\n",
       "                                 number_of_events_passing_fitc_gate  \\\n",
       "job_id    specimen     tube                                           \n",
       "Job_60359 Specimen_001 Tube_006                               21064   \n",
       "                       Tube_007                                6140   \n",
       "                       Tube_008                               19489   \n",
       "                       Tube_009                                2217   \n",
       "Job_60063 Specimen_001 Tube_006                               64106   \n",
       "                       Tube_007                               52457   \n",
       "                       Tube_008                               64109   \n",
       "                       Tube_009                               53023   \n",
       "Job_59787 Specimen_001 Tube_006                               37520   \n",
       "                       Tube_007                               32048   \n",
       "                       Tube_008                               29746   \n",
       "                       Tube_009                               30930   \n",
       "                       Tube_010                               27904   \n",
       "\n",
       "                                 fraction_collected  \n",
       "job_id    specimen     tube                          \n",
       "Job_60359 Specimen_001 Tube_006            0.210648  \n",
       "                       Tube_007            0.061403  \n",
       "                       Tube_008            0.194896  \n",
       "                       Tube_009            0.022171  \n",
       "Job_60063 Specimen_001 Tube_006            0.652127  \n",
       "                       Tube_007            0.524580  \n",
       "                       Tube_008            0.641135  \n",
       "                       Tube_009            0.530246  \n",
       "Job_59787 Specimen_001 Tube_006            0.375211  \n",
       "                       Tube_007            0.320483  \n",
       "                       Tube_008            0.297460  \n",
       "                       Tube_009            0.309303  \n",
       "                       Tube_010            0.279043  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "map_job_id_to_sort_round = {\n",
    "    'Job_59787' : 1, 'Job_60063' : 2, 'Job_60359' : 3\n",
    "}\n",
    "naive_sample_expression = {'trypsin':{}, 'chymotrypsin':{}}\n",
    "list_of_indices_to_drop = []\n",
    "\n",
    "for (i, row) in manifest_df.iterrows():\n",
    "    (job_id, specimen, tube) = i\n",
    "    sort_round = map_job_id_to_sort_round[job_id]\n",
    "    if row['concentration'] == 0.0:\n",
    "        naive_sample_expression[row['protease']][sort_round] = row['fraction_collected']\n",
    "        if sort_round != 1:\n",
    "            list_of_indices_to_drop.append(i)\n",
    "            \n",
    "manifest_df.drop(list_of_indices_to_drop, axis='index', inplace=True)\n",
    "manifest_df[columns_to_show]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in data from the XML files on the total number of cells collected. I will do this for all XML files the BIOFAB has uploaded, even ones that aren't relevant to this experiment. Then, I will identify the relevant ones using the `job_id`, `specimen`, and `tube` columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Compile data in each XML file\n",
    "xml_data_dict = {\n",
    "    key : []\n",
    "    for key in ['job_id', 'specimen', 'tube', 'cells_collected', 'intended_number_of_cells_collected'] #\n",
    "}\n",
    "cells_collected_pattern = re.compile(r'\\w+ : (?P<cells_collected>\\d+) / (?P<intended_number_of_cells_collected>\\d+)')\n",
    "for filename in xml_files:\n",
    "    tree = ET.parse(filename)\n",
    "    root = tree.getroot()\n",
    "\n",
    "    # Get experiment metadata\n",
    "    xml_data_dict['job_id'].append( root.findall(\".//*[@name='Experiment']\")[0].get('value') )\n",
    "    xml_data_dict['specimen'].append( root.findall(\".//*[@name='Specimen']\")[0].get('value') )\n",
    "    xml_data_dict['tube'].append( root.findall(\".//*[@name='Tube']\")[0].get('value') )\n",
    "    xml_data_dict['xmlfile'] = filename\n",
    "    \n",
    "    # Get experiment counts\n",
    "    cells_collected_data = root[4].findall(\".//*[@col='1']\")[0].text\n",
    "    match = re.match(cells_collected_pattern, cells_collected_data)\n",
    "    if match:\n",
    "        xml_data_dict['cells_collected'].append(\n",
    "            int(match.group('cells_collected'))\n",
    "        )\n",
    "        xml_data_dict['intended_number_of_cells_collected'].append(\n",
    "            int(match.group('intended_number_of_cells_collected'))\n",
    "        )\n",
    "    else:\n",
    "        xml_data_dict['cells_collected'].append(None)\n",
    "        xml_data_dict['intended_number_of_cells_collected'].append(None)\n",
    "          \n",
    "# Convert data to a dataframe\n",
    "xml_data_df = pandas.DataFrame.from_dict(xml_data_dict)\n",
    "xml_data_df.set_index(['job_id', 'specimen', 'tube'], inplace=True)\n",
    "#xml_data_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Merge the FACS data from the two input sources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>cells_collected</th>\n",
       "      <th>fraction_collected</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protease</th>\n",
       "      <th>concentration</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">chymotrypsin</th>\n",
       "      <th>9.5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.320483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28.0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.297460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83.0</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>0.652127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250.0</th>\n",
       "      <td>6312000.0</td>\n",
       "      <td>0.524580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750.0</th>\n",
       "      <td>2107594.0</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750.0</th>\n",
       "      <td>3954.0</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2250.0</th>\n",
       "      <td>648405.0</td>\n",
       "      <td>0.061403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">trypsin</th>\n",
       "      <th>0.0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9.5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.309303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28.0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.279043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83.0</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>0.641135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250.0</th>\n",
       "      <td>6384000.0</td>\n",
       "      <td>0.530246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>750.0</th>\n",
       "      <td>2137779.0</td>\n",
       "      <td>0.194896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2250.0</th>\n",
       "      <td>240000.0</td>\n",
       "      <td>0.022171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            cells_collected  fraction_collected\n",
       "protease     concentration                                     \n",
       "chymotrypsin 9.5                        NaN            0.320483\n",
       "             28.0                       NaN            0.297460\n",
       "             83.0                 7704000.0            0.652127\n",
       "             250.0                6312000.0            0.524580\n",
       "             750.0                2107594.0            0.210648\n",
       "             750.0                   3954.0            0.210648\n",
       "             2250.0                648405.0            0.061403\n",
       "trypsin      0.0                        NaN            0.375211\n",
       "             9.5                        NaN            0.309303\n",
       "             28.0                       NaN            0.279043\n",
       "             83.0                 7704000.0            0.641135\n",
       "             250.0                6384000.0            0.530246\n",
       "             750.0                2137779.0            0.194896\n",
       "             2250.0                240000.0            0.022171"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "facs_df = manifest_df.merge(xml_data_df, left_index=True, right_index=True, how='left')\n",
    "facs_df = facs_df[\n",
    "    facs_df['strain'] == library_name\n",
    "]\n",
    "facs_df.set_index(['protease', 'concentration'], inplace=True, drop=False)\n",
    "columns_to_show = ['cells_collected', 'fraction_collected']\n",
    "facs_df[columns_to_show].sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Merge the FACS and deep-sequencing data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protease</th>\n",
       "      <th>concentration</th>\n",
       "      <th>fastq_id</th>\n",
       "      <th>cells_collected</th>\n",
       "      <th>fraction_collected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128444/Files/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.320483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128446/Files/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.297460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128456/Files/</td>\n",
       "      <td>7704000.0</td>\n",
       "      <td>0.652127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128452/Files/</td>\n",
       "      <td>6312000.0</td>\n",
       "      <td>0.524580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128460/Files/</td>\n",
       "      <td>2107594.0</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128460/Files/</td>\n",
       "      <td>3954.0</td>\n",
       "      <td>0.210648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128462/Files/</td>\n",
       "      <td>648405.0</td>\n",
       "      <td>0.061403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128448/Files/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.309303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128450/Files/</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.279043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128454/Files/</td>\n",
       "      <td>7704000.0</td>\n",
       "      <td>0.641135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128458/Files/</td>\n",
       "      <td>6384000.0</td>\n",
       "      <td>0.530246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128464/Files/</td>\n",
       "      <td>2137779.0</td>\n",
       "      <td>0.194896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>trypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128466/Files/</td>\n",
       "      <td>240000.0</td>\n",
       "      <td>0.022171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        protease  concentration       fastq_id  cells_collected  \\\n",
       "0   chymotrypsin            9.5  128444/Files/              NaN   \n",
       "1   chymotrypsin           28.0  128446/Files/              NaN   \n",
       "2   chymotrypsin           83.0  128456/Files/        7704000.0   \n",
       "3   chymotrypsin          250.0  128452/Files/        6312000.0   \n",
       "4   chymotrypsin          750.0  128460/Files/        2107594.0   \n",
       "5   chymotrypsin          750.0  128460/Files/           3954.0   \n",
       "6   chymotrypsin         2250.0  128462/Files/         648405.0   \n",
       "7        trypsin            0.0  128412/Files/              NaN   \n",
       "8        trypsin            9.5  128448/Files/              NaN   \n",
       "9        trypsin           28.0  128450/Files/              NaN   \n",
       "10       trypsin           83.0  128454/Files/        7704000.0   \n",
       "11       trypsin          250.0  128458/Files/        6384000.0   \n",
       "12       trypsin          750.0  128464/Files/        2137779.0   \n",
       "13       trypsin         2250.0  128466/Files/         240000.0   \n",
       "\n",
       "    fraction_collected  \n",
       "0             0.320483  \n",
       "1             0.297460  \n",
       "2             0.652127  \n",
       "3             0.524580  \n",
       "4             0.210648  \n",
       "5             0.210648  \n",
       "6             0.061403  \n",
       "7             0.375211  \n",
       "8             0.309303  \n",
       "9             0.279043  \n",
       "10            0.641135  \n",
       "11            0.530246  \n",
       "12            0.194896  \n",
       "13            0.022171  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = fastq_df.merge(facs_df, left_index=True, right_index=True, how=\"outer\")\n",
    "columns_to_write = ['fastq_id', 'cells_collected', 'fraction_collected']\n",
    "df = df[columns_to_write]\n",
    "df.reset_index(inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are two entries for (chymotrypsin, 750.0) one with a large number of counts, one with a small number. I will remove the row with the smaller number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df.drop([5], inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add a row for a naive sample for chymotrypsin, using the same deep-sequencing data as for trypsin "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_with_extra_row = pandas.DataFrame.from_dict(\n",
    "    {\n",
    "        'protease' : ['chymotrypsin'],\n",
    "        'concentration' : [0],\n",
    "        'fastq_id' : [df.iloc[6]['fastq_id']],\n",
    "        'cells_collected' : [df.iloc[6]['cells_collected']],\n",
    "        'fraction_collected' : [df.iloc[6]['fraction_collected']]\n",
    "    }\n",
    ")\n",
    "df_with_extra_row\n",
    "df = pandas.concat([df, df_with_extra_row])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show concatenated dataframe, sorted by protease and protease concentration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cells_collected</th>\n",
       "      <th>concentration</th>\n",
       "      <th>fastq_id</th>\n",
       "      <th>fraction_collected</th>\n",
       "      <th>protease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128444/Files/</td>\n",
       "      <td>0.320483</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128446/Files/</td>\n",
       "      <td>0.297460</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128456/Files/</td>\n",
       "      <td>0.652127</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6312000.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128452/Files/</td>\n",
       "      <td>0.524580</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2107594.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128460/Files/</td>\n",
       "      <td>0.210648</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>648405.0</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128462/Files/</td>\n",
       "      <td>0.061403</td>\n",
       "      <td>chymotrypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128448/Files/</td>\n",
       "      <td>0.309303</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128450/Files/</td>\n",
       "      <td>0.279043</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128454/Files/</td>\n",
       "      <td>0.641135</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>6384000.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128458/Files/</td>\n",
       "      <td>0.530246</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2137779.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128464/Files/</td>\n",
       "      <td>0.194896</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>240000.0</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128466/Files/</td>\n",
       "      <td>0.022171</td>\n",
       "      <td>trypsin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cells_collected  concentration       fastq_id  fraction_collected  \\\n",
       "0               NaN            0.0  128412/Files/            0.375211   \n",
       "0               NaN            9.5  128444/Files/            0.320483   \n",
       "1               NaN           28.0  128446/Files/            0.297460   \n",
       "2         7704000.0           83.0  128456/Files/            0.652127   \n",
       "3         6312000.0          250.0  128452/Files/            0.524580   \n",
       "4         2107594.0          750.0  128460/Files/            0.210648   \n",
       "6          648405.0         2250.0  128462/Files/            0.061403   \n",
       "7               NaN            0.0  128412/Files/            0.375211   \n",
       "8               NaN            9.5  128448/Files/            0.309303   \n",
       "9               NaN           28.0  128450/Files/            0.279043   \n",
       "10        7704000.0           83.0  128454/Files/            0.641135   \n",
       "11        6384000.0          250.0  128458/Files/            0.530246   \n",
       "12        2137779.0          750.0  128464/Files/            0.194896   \n",
       "13         240000.0         2250.0  128466/Files/            0.022171   \n",
       "\n",
       "        protease  \n",
       "0   chymotrypsin  \n",
       "0   chymotrypsin  \n",
       "1   chymotrypsin  \n",
       "2   chymotrypsin  \n",
       "3   chymotrypsin  \n",
       "4   chymotrypsin  \n",
       "6   chymotrypsin  \n",
       "7        trypsin  \n",
       "8        trypsin  \n",
       "9        trypsin  \n",
       "10       trypsin  \n",
       "11       trypsin  \n",
       "12       trypsin  \n",
       "13       trypsin  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.sort_values(by=['protease', 'concentration'], inplace=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually add missing columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Add columns\n",
    "df['experiment_id'] = library_name\n",
    "df['selection_strength'] = 2 * [i for i in range(0,7)]\n",
    "df['conc_factor'] = 2 * ['', '3', '3', '3', '3', '3', '3']\n",
    "df['parent'] = 2 * ['', '0', '0', '2', '2', '4', '4']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the row called `parent_expression`, which reports the fraction of collected cells (i.e., passed the selection threshold) in a naive sample that was ***not*** treated with protease and which was sorted on the same day as protease-treated samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cells_collected</th>\n",
       "      <th>concentration</th>\n",
       "      <th>fastq_id</th>\n",
       "      <th>fraction_collected</th>\n",
       "      <th>protease</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>selection_strength</th>\n",
       "      <th>conc_factor</th>\n",
       "      <th>parent</th>\n",
       "      <th>parent_expression</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128444/Files/</td>\n",
       "      <td>0.320483</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128446/Files/</td>\n",
       "      <td>0.297460</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128456/Files/</td>\n",
       "      <td>0.652127</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.678084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6312000.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128452/Files/</td>\n",
       "      <td>0.524580</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.678084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2107594.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128460/Files/</td>\n",
       "      <td>0.210648</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.775654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>648405.0</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128462/Files/</td>\n",
       "      <td>0.061403</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.775654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>128448/Files/</td>\n",
       "      <td>0.309303</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>128450/Files/</td>\n",
       "      <td>0.279043</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0.375211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7704000.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>128454/Files/</td>\n",
       "      <td>0.641135</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.704705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>6384000.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>128458/Files/</td>\n",
       "      <td>0.530246</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.704705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2137779.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>128464/Files/</td>\n",
       "      <td>0.194896</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.762613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>240000.0</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>128466/Files/</td>\n",
       "      <td>0.022171</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0.762613</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    cells_collected  concentration       fastq_id  fraction_collected  \\\n",
       "0               NaN            0.0  128412/Files/            0.375211   \n",
       "0               NaN            9.5  128444/Files/            0.320483   \n",
       "1               NaN           28.0  128446/Files/            0.297460   \n",
       "2         7704000.0           83.0  128456/Files/            0.652127   \n",
       "3         6312000.0          250.0  128452/Files/            0.524580   \n",
       "4         2107594.0          750.0  128460/Files/            0.210648   \n",
       "6          648405.0         2250.0  128462/Files/            0.061403   \n",
       "7               NaN            0.0  128412/Files/            0.375211   \n",
       "8               NaN            9.5  128448/Files/            0.309303   \n",
       "9               NaN           28.0  128450/Files/            0.279043   \n",
       "10        7704000.0           83.0  128454/Files/            0.641135   \n",
       "11        6384000.0          250.0  128458/Files/            0.530246   \n",
       "12        2137779.0          750.0  128464/Files/            0.194896   \n",
       "13         240000.0         2250.0  128466/Files/            0.022171   \n",
       "\n",
       "        protease experiment_id  selection_strength conc_factor parent  \\\n",
       "0   chymotrypsin     EVA_NOV17                   0                      \n",
       "0   chymotrypsin     EVA_NOV17                   1           3      0   \n",
       "1   chymotrypsin     EVA_NOV17                   2           3      0   \n",
       "2   chymotrypsin     EVA_NOV17                   3           3      2   \n",
       "3   chymotrypsin     EVA_NOV17                   4           3      2   \n",
       "4   chymotrypsin     EVA_NOV17                   5           3      4   \n",
       "6   chymotrypsin     EVA_NOV17                   6           3      4   \n",
       "7        trypsin     EVA_NOV17                   0                      \n",
       "8        trypsin     EVA_NOV17                   1           3      0   \n",
       "9        trypsin     EVA_NOV17                   2           3      0   \n",
       "10       trypsin     EVA_NOV17                   3           3      2   \n",
       "11       trypsin     EVA_NOV17                   4           3      2   \n",
       "12       trypsin     EVA_NOV17                   5           3      4   \n",
       "13       trypsin     EVA_NOV17                   6           3      4   \n",
       "\n",
       "   parent_expression  \n",
       "0                     \n",
       "0           0.375211  \n",
       "1           0.375211  \n",
       "2           0.678084  \n",
       "3           0.678084  \n",
       "4           0.775654  \n",
       "6           0.775654  \n",
       "7                     \n",
       "8           0.375211  \n",
       "9           0.375211  \n",
       "10          0.704705  \n",
       "11          0.704705  \n",
       "12          0.762613  \n",
       "13          0.762613  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a dictionary to map protease concentration to sort round\n",
    "map_protease_conc_to_sort_round = {\n",
    "    0.0 : 1,\n",
    "    9.5 : 1,\n",
    "    28.0 : 1,\n",
    "    83.0 : 2,\n",
    "    250.0 : 2,\n",
    "    750.0 : 3,\n",
    "    2250.0 : 3\n",
    "}\n",
    "\n",
    "# Add a `parent_expression` column to the dataframe. Use the same value for\n",
    "# all protease-treated samples from the first sort round, regardless of protease.\n",
    "# This value is derived from (trypsin, 0), but is relative to both proteases\n",
    "# since it is naive sample. Add blank entires for the naive samples from the first\n",
    "# sort.\n",
    "parent_expression_list = []\n",
    "for (i, row) in df.iterrows():\n",
    "    protease = row['protease']\n",
    "    concentration = row['concentration']\n",
    "    sort_round = map_protease_conc_to_sort_round[concentration]\n",
    "    if concentration == 0.0:\n",
    "        parent_expression_list.append('')\n",
    "    elif (protease, sort_round) == ('chymotrypsin', 1):\n",
    "        parent_expression_list.append(naive_sample_expression['trypsin'][sort_round])\n",
    "    else:\n",
    "        parent_expression_list.append(naive_sample_expression[protease][sort_round])\n",
    "df['parent_expression'] = parent_expression_list\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename columns and show the final dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>protease_type</th>\n",
       "      <th>concentration</th>\n",
       "      <th>selection_strength</th>\n",
       "      <th>parent</th>\n",
       "      <th>conc_factor</th>\n",
       "      <th>fastq_id</th>\n",
       "      <th>parent_expression</th>\n",
       "      <th>fraction_collected</th>\n",
       "      <th>cells_collected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td></td>\n",
       "      <td>0.375211</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>128444/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>0.320483</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>128446/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>0.297460</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>128456/Files/</td>\n",
       "      <td>0.678084</td>\n",
       "      <td>0.652127</td>\n",
       "      <td>7704000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>128452/Files/</td>\n",
       "      <td>0.678084</td>\n",
       "      <td>0.524580</td>\n",
       "      <td>6312000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>128460/Files/</td>\n",
       "      <td>0.775654</td>\n",
       "      <td>0.210648</td>\n",
       "      <td>2107594.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>chymotrypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>128462/Files/</td>\n",
       "      <td>0.775654</td>\n",
       "      <td>0.061403</td>\n",
       "      <td>648405.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>128412/Files/</td>\n",
       "      <td></td>\n",
       "      <td>0.375211</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>9.5</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>128448/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>0.309303</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>28.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>128450/Files/</td>\n",
       "      <td>0.375211</td>\n",
       "      <td>0.279043</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>83.0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>128454/Files/</td>\n",
       "      <td>0.704705</td>\n",
       "      <td>0.641135</td>\n",
       "      <td>7704000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>250.0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>128458/Files/</td>\n",
       "      <td>0.704705</td>\n",
       "      <td>0.530246</td>\n",
       "      <td>6384000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>750.0</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>128464/Files/</td>\n",
       "      <td>0.762613</td>\n",
       "      <td>0.194896</td>\n",
       "      <td>2137779.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EVA_NOV17</td>\n",
       "      <td>trypsin</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>128466/Files/</td>\n",
       "      <td>0.762613</td>\n",
       "      <td>0.022171</td>\n",
       "      <td>240000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   experiment_id protease_type  concentration  selection_strength parent  \\\n",
       "0      EVA_NOV17  chymotrypsin            0.0                   0          \n",
       "0      EVA_NOV17  chymotrypsin            9.5                   1      0   \n",
       "1      EVA_NOV17  chymotrypsin           28.0                   2      0   \n",
       "2      EVA_NOV17  chymotrypsin           83.0                   3      2   \n",
       "3      EVA_NOV17  chymotrypsin          250.0                   4      2   \n",
       "4      EVA_NOV17  chymotrypsin          750.0                   5      4   \n",
       "6      EVA_NOV17  chymotrypsin         2250.0                   6      4   \n",
       "7      EVA_NOV17       trypsin            0.0                   0          \n",
       "8      EVA_NOV17       trypsin            9.5                   1      0   \n",
       "9      EVA_NOV17       trypsin           28.0                   2      0   \n",
       "10     EVA_NOV17       trypsin           83.0                   3      2   \n",
       "11     EVA_NOV17       trypsin          250.0                   4      2   \n",
       "12     EVA_NOV17       trypsin          750.0                   5      4   \n",
       "13     EVA_NOV17       trypsin         2250.0                   6      4   \n",
       "\n",
       "   conc_factor       fastq_id parent_expression  fraction_collected  \\\n",
       "0               128412/Files/                              0.375211   \n",
       "0            3  128444/Files/          0.375211            0.320483   \n",
       "1            3  128446/Files/          0.375211            0.297460   \n",
       "2            3  128456/Files/          0.678084            0.652127   \n",
       "3            3  128452/Files/          0.678084            0.524580   \n",
       "4            3  128460/Files/          0.775654            0.210648   \n",
       "6            3  128462/Files/          0.775654            0.061403   \n",
       "7               128412/Files/                              0.375211   \n",
       "8            3  128448/Files/          0.375211            0.309303   \n",
       "9            3  128450/Files/          0.375211            0.279043   \n",
       "10           3  128454/Files/          0.704705            0.641135   \n",
       "11           3  128458/Files/          0.704705            0.530246   \n",
       "12           3  128464/Files/          0.762613            0.194896   \n",
       "13           3  128466/Files/          0.762613            0.022171   \n",
       "\n",
       "    cells_collected  \n",
       "0               NaN  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2         7704000.0  \n",
       "3         6312000.0  \n",
       "4         2107594.0  \n",
       "6          648405.0  \n",
       "7               NaN  \n",
       "8               NaN  \n",
       "9               NaN  \n",
       "10        7704000.0  \n",
       "11        6384000.0  \n",
       "12        2137779.0  \n",
       "13         240000.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename columns\n",
    "df.rename(\n",
    "    {'protease':'protease_type'},\n",
    "    inplace=True,\n",
    "    axis='columns'\n",
    ")\n",
    "column_order = [\n",
    "    'experiment_id', 'protease_type', 'concentration', 'selection_strength', 'parent', 'conc_factor', 'fastq_id',\n",
    "    'parent_expression', 'fraction_collected', 'cells_collected'\n",
    "]\n",
    "df[column_order]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write the above dataframe to an output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Writing metadata to the experimental summary file: results/171201_Eva_10861/parse_metadata/experimental_metadata_from_script.csv\n"
     ]
    }
   ],
   "source": [
    "print(\"Writing metadata to the experimental summary file: {0}\".format(experimental_metadata_output_file))\n",
    "df[column_order].to_csv(experimental_metadata_output_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Compute EC50 values from the input deep-sequencing and FACS data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'experimental_metadata_output_file' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-4e51c04e8770>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[0mdata_dir\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'data/171201_Eva/'\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mdesigned_sequences_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mos\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata_dir\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'designed_sequences.csv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m \u001b[0mexperimental_summary_file\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mexperimental_metadata_output_file\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[0mfastq_dir\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mfastq_dir\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mpear_path\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'/home/05402/haddox/software/pear/bin/pear'\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'experimental_metadata_output_file' is not defined"
     ]
    }
   ],
   "source": [
    "# Define input variables\n",
    "data_dir = 'data/171201_Eva/'\n",
    "designed_sequences_file = os.path.join(data_dir, 'designed_sequences.csv')\n",
    "experimental_summary_file = experimental_metadata_output_file\n",
    "fastq_dir = fastq_dir\n",
    "pear_path = '/home/05402/haddox/software/pear/bin/pear'\n",
    "output_dir = resultsdir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "python scripts/compute_ec50_values_from_deep_sequencing_data.py --designed_sequences_file data/171201_Eva/designed_sequences.csv --experimental_summary_file results/171201_Eva_10861/parse_metadata/experimental_metadata_from_script.csv --fastq_dir /work/05402/haddox/jupyter/sd2e-community/shared-q1-workshop/strcklnd/ngs_data/run_8/EVA_Nov17_new_prep/ --pear_path /home/05402/haddox/software/pear/bin/pear --output_dir results/171201_Eva_10861\n"
     ]
    }
   ],
   "source": [
    "# Write the command to carry everything out\n",
    "cmd = ' '.join([\n",
    "    'python',\n",
    "    'scripts/compute_ec50_values_from_deep_sequencing_data.py',\n",
    "    '--designed_sequences_file {0}'.format(designed_sequences_file),\n",
    "    '--experimental_summary_file {0}'.format(experimental_summary_file),\n",
    "    '--fastq_dir {0}'.format(fastq_dir),\n",
    "    '--pear_path {0}'.format(pear_path),\n",
    "    '--output_dir {0}'.format(output_dir)\n",
    "])\n",
    "\n",
    "#! {cmd}\n",
    "print(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze the stability scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To do:\n",
    "\n",
    "* Quantify sequencing depth\n",
    "* Look at the distribution of counts over time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quantify the number of assembled reads per sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ParsePAREOutfile(outfile):\n",
    "    \"\"\"\n",
    "    This function parses the output data generated by PARE when assemblying paired-end reads\n",
    "    \n",
    "    Args:\n",
    "        `outfile`: the path to a file with the output data generated by PARE\n",
    "        \n",
    "    Returns:\n",
    "        A tupple with the following three variables in the order they appear in the below list:\n",
    "            `n_assembled_reads` : the total number of assembled reads\n",
    "            `n_discarded_reads` : the total number of discarded reads\n",
    "            `n_non_assembled_reads` : the total number of non_assembled_reads\n",
    "    \"\"\"\n",
    "    \n",
    "    # Pattern used to extract data\n",
    "    n_reads_pattern = re.compile(r'\\: (?P<n_reads>[\\d,]+) /')\n",
    "    \n",
    "    # Use regular expressions to extract the relevant info from the file\n",
    "    n_assembled_reads = n_discarded_reads = n_non_assembled_reads = n_total_reads = None\n",
    "    with open(outfile) as f:\n",
    "        #print(f.readlines())\n",
    "        for line in f:\n",
    "            if 'Assembled reads .' in line:\n",
    "                if n_assembled_reads:\n",
    "                    raise ValueError(\"Already found data for `n_assembled_reads`\")\n",
    "                n_assembled_match = re.search(n_reads_pattern, line)\n",
    "                n_assembled_reads = int(n_assembled_match.group('n_reads').replace(',', ''))\n",
    "                \n",
    "            elif 'Discarded reads .' in line:\n",
    "                if n_discarded_reads:\n",
    "                    raise ValueError(\"Already found data for `n_discarded_reads`\")\n",
    "                n_discarded_match = re.search(n_reads_pattern, line)\n",
    "                n_discarded_reads = int(n_discarded_match.group('n_reads').replace(',', ''))\n",
    "                \n",
    "            elif 'Not assembled reads .' in line:\n",
    "                if n_non_assembled_reads:\n",
    "                    raise ValueError(\"Already found data for `n_non_assembled_reads`\")\n",
    "                n_non_assembled_match = re.search(n_reads_pattern, line)\n",
    "                n_non_assembled_reads = int(n_non_assembled_match.group('n_reads').replace(',', ''))\n",
    "    \n",
    "    return (n_assembled_reads, n_discarded_reads, n_non_assembled_reads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Find log files from PARE\n",
    "pare_log_dir = os.path.join(resultsdir, 'paired_FASTQ_files')\n",
    "\n",
    "selection_indices = list(range(0,7))\n",
    "experiment_names = ['trypsin_{0}'.format(i) for i in selection_indices] + \\\n",
    "    ['chymotrypsin_{0}'.format(i) for i in selection_indices]\n",
    "log_files = glob.glob(os.path.join(pare_log_dir, '*.log'))\n",
    "log_files_dict = {\n",
    "    experiment_name : []\n",
    "    for experiment_name in experiment_names\n",
    "}\n",
    "# For each log file, look for an experimental name that matches the beginning\n",
    "# of the base name of the log file\n",
    "for log_file_name in log_files:\n",
    "    matching_experiments = []\n",
    "    for experiment_name in log_files_dict:\n",
    "        log_file_basename = os.path.basename(log_file_name)\n",
    "        if log_file_basename.find('{0}-'.format(experiment_name)) == 0:\n",
    "            log_files_dict[experiment_name].append(log_file_name)\n",
    "            matching_experiments.append(experiment_name)\n",
    "    if len(matching_experiments) == 0:\n",
    "        raise ValueError(\"Could not find a matching experiment for the log file: {0}\".format(log_file_name))\n",
    "    if len(matching_experiments) > 1:\n",
    "        print(matching_experiments)\n",
    "        raise ValueError(\"Found multiple matching experiments for the log file: {0}\".format(log_file_name))\n",
    "    assert len(matching_experiments) == 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read in depth and quality scores from the output of the `PEAR` program and store in a pandas dataframe\n",
    "assembly_d = {\n",
    "    key : []\n",
    "    for key in ['experiment_name', 'n_assembled_reads', 'n_discarded_reads', 'n_non_assembled_reads']\n",
    "}\n",
    "for experiment_name in log_files_dict:\n",
    "    assembly_d['experiment_name'].append(experiment_name)\n",
    "    n_assembled_reads = n_discarded_reads = n_non_assembled_reads = 0\n",
    "    for log_file_name in log_files_dict[experiment_name]:\n",
    "        (n_assembled_reads_i, n_discarded_reads_i, n_non_assembled_reads_i) = \\\n",
    "            ParsePAREOutfile(log_file_name)\n",
    "        n_assembled_reads += n_assembled_reads_i\n",
    "        n_discarded_reads += n_discarded_reads_i\n",
    "        n_non_assembled_reads += n_non_assembled_reads_i\n",
    "    assembly_d['n_assembled_reads'].append(n_assembled_reads)\n",
    "    assembly_d['n_discarded_reads'].append(n_discarded_reads)\n",
    "    assembly_d['n_non_assembled_reads'].append(n_non_assembled_reads)\n",
    "\n",
    "assembly_df = pandas.DataFrame.from_dict(assembly_d)\n",
    "\n",
    "# Plot the data for each replicate as stacked bar charts\n",
    "fig = plt.figure(figsize=(10,7))\n",
    "\n",
    "# Get data for each bar\n",
    "assembly_df.set_index('experiment_name', inplace=True)\n",
    "labels = sorted(assembly_df.index.values)\n",
    "first_bar = assembly_df.loc[labels]['n_assembled_reads']\n",
    "second_bar = assembly_df.loc[labels]['n_non_assembled_reads']\n",
    "third_bar = assembly_df.loc[labels]['n_discarded_reads']\n",
    "assert(len(first_bar) == len(second_bar))\n",
    "\n",
    "# Make plot\n",
    "plot_indices = np.arange(len(first_bar))\n",
    "width = 0.35\n",
    "plt.barh(plot_indices, first_bar, label='assembled', align='center')\n",
    "plt.barh(plot_indices, second_bar, left=first_bar, label='non-assembled', color='red', align='center')\n",
    "plt.barh(plot_indices, third_bar, left=second_bar, label='discarded', color='purple', align='center')\n",
    "plt.yticks(plot_indices, labels)\n",
    "plt.xlabel('Number of reads')\n",
    "plt.legend()\n",
    "plt.ticklabel_format(style='sci', axis='x', scilimits=(0,0))\n",
    "def adjust_ylim(l):\n",
    "    mn, mx = l\n",
    "    mn += 1.5\n",
    "    mx -= .5\n",
    "    return (mn, mx)\n",
    "plt.ylim(adjust_ylim(plt.ylim()))\n",
    "plt.yticks()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the distribution of stability scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trypsin_df = pandas.read_csv('results/Inna_April_2016/stability_scores/trypsin_stability_scores.txt', sep='\\t')\n",
    "chymotrypsin_df = pandas.read_csv('results/Inna_April_2016/stability_scores/chymotrypsin_stability_scores.txt', sep='\\t')\n",
    "sns.distplot(trypsin_df['stabilityscore'])\n",
    "sns.distplot(chymotrypsin_df['stabilityscore'])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "trypsin_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import gzip\n",
    "\n",
    "filename = '/work/05402/haddox/jupyter/sd2e-community/shared-q1-workshop/strcklnd/ngs_data/run_6/EVA_Nov17/125459/Files/125459-naive_S1_L001_R1_001.fastq.gz'\n",
    "with gzip.open(filename, 'rb') as f:\n",
    "    for (i, line) in enumerate(f):\n",
    "        print(line.find('CATATG'))\n",
    "        print(line)\n",
    "        if i > 100:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Truncate the length of input design sequences to match the length of the deep-sequencing reads\n",
    "\n",
    "The sequence `CATATG` appears at about nucleotide 24-27 in the forward reads. Since the forward reads are only 150 nt, and since the ends of the first ten reads look low quality, I will trim the input design sequences to the first 30 amino acids (=90/3)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = pandas.read_csv('data/171201_Eva/designed_sequences.csv')\n",
    "df.rename(\n",
    "    {'protein_sequence':'full_sequence'},\n",
    "    inplace=True,\n",
    "    axis='columns'\n",
    ")\n",
    "df['protein_sequence'] = df['full_sequence'].apply(\n",
    "    lambda x: str(x)[:30]\n",
    ")\n",
    "\n",
    "shortened_designed_sequences_file = os.path.join(\n",
    "    resultsdir_parse_metadata,\n",
    "    'shortened_designed_sequences.csv'\n",
    ")\n",
    "print(\"Writing shortened design sequences to the file: {0}\".format(\n",
    "    shortened_designed_sequences_file\n",
    "))\n",
    "df[['name', 'protein_sequence']].to_csv(\n",
    "    shortened_designed_sequences_file, index=False\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
